include/boost/random/faure.hpp

```diff
@@ -0,0 +1,363 @@
+
+  packed_uint_t& upper_element(size_type i, size_type j, size_type dim)
+  {
+    BOOST_ASSERT( i < dim );
+    BOOST_ASSERT( j < dim );
+    BOOST_ASSERT( i <= j );
+    return coeff[(i * (2 * dim - i + 1)) / 2 + j - i];
+  }
+
+  template<typename Iterator>
+  RealType compute_recip(size_type seq, Iterator out) const
+  {
+    // Here we do
+    //   Sum ( 0 <= J <= HISUM ) YTEMP(J) * QS**J
+    //   Sum ( 0 <= J <= HISUM ) YTEMP(J) / QS**(J+1)
+    // in one go
+    RealType r = RealType();
+    size_type m, k = qs_pow;
+    for( ; k != 0; ++out, seq = m, k /= qs_base )
+    {
+      m  = seq % k;
+      RealType v  = (seq - m) / k; // RealType <- size type
+      r += v;
+      r /= static_cast<RealType>(qs_base);
+      *out = v; // saves double dereference
+    }
+    return r;
+  }
+
+  void compute_coefficients(const size_type n)
+  {
+    // Resize and initialize to zero
+    coeff.resize(size_hint(n));
+    std::fill(coeff.begin(), coeff.end(), packed_uint_t());
+
+    // The first row and the diagonal is assigned to 1
+    upper_element(0, 0, n) = 1;
+    for (size_type i = 1; i < n; ++i)
+    {
+      upper_element(0, i, n) = 1;
+      upper_element(i, i, n) = 1;
+    }
+
+    // Computes binomial coefficients MOD qs_base
+    for (size_type i = 1; i < n; ++i)
+    {
+      for (size_type j = i + 1; j < n; ++j)
+      {
+        upper_element(i, j, n) = ( upper_element(i, j-1, n) +
+                                   upper_element(i-1, j-1, n) ) % qs_base;
+      }
+    }
+  }
+
+private:
+  packed_uint_t qs_base;
+
+  // here we cache precomputed data; note that binomial coefficients have
+  // to be recomputed iff the integer part of the logarithm of seq changes,
+  // which happens relatively rarely.
+  std::vector<packed_uint_t> coeff; // packed upper (!) triangular matrix
+  std::vector<RealType> ytemp;
+  size_type qs_pow;
+};
+
+}} // namespace detail::fr
+
+typedef detail::qrng_tables::primes default_faure_prime_table;
+
+/** @endcond */
+
+//!Instantiations of class template faure_engine model a \quasi_random_number_generator.
+//!The faure_engine uses the algorithm described in
+//! \blockquote
+//!Henri Faure,
+//!Discrepance de suites associees a un systeme de numeration (en dimension s),
+//!Acta Arithmetica,
+//!Volume 41, 1982, pages 337-351.
+//! \endblockquote
+//
+//! \blockquote
+//!Bennett Fox,
+//!Algorithm 647:
+//!Implementation and Relative Efficiency of Quasirandom
+//!Sequence Generators,
+//!ACM Transactions on Mathematical Software,
+//!Volume 12, Number 4, December 1986, pages 362-376.
+//! \endblockquote
+//!
+//!In the following documentation @c X denotes the concrete class of the template
+//!faure_engine returning objects of type @c RealType, u and v are the values of @c X.
+//!
+//!Some member functions may throw exceptions of type @c std::bad_alloc.
+template<typename RealType, typename SeqSizeT, typename PrimeTable = default_faure_prime_table>
+class faure_engine
+  : public detail::qrng_base<
```

Comment:
This has the unfortunate side-effect that boost::random::detail is an associated namespace of faure_engine.  I generally avoid putting internal base classes in namespace detail.  (You can also put qrng_base in its own namespace).

Reply:
I see, I see... Would it be acceptable to deepen the namespace to, say, detail::quasi::qrng_base?
IIRC, ADL would get into effect only for the innermost enclosing namespaces, no?

Reply:
I think this is done now. I moved qrng_base, gray_coded_qrng from namespace detail to qrng_detail.